### ISSUE TITLE
DAQModule design encourages race between commands and worker thread

### ISSUE THREAD

**Issue by brettviren** [DUNE-DAQ/appfwk#118]
DAQModule as a base does not address how one or more "worker threads" coordinate with the main thread that services commands through registered handler mehods.

In fact, it is possible to deliver a command train: init, conf, start, stop before the worker thread even gets started.  When that `do_work()` method finally gets called, the module is very confused.  In tests with the `Fake*` modules this leads to either no work, or a 30s hang followed by a crash.

This particular race is lost due to very fast command delivery.  This fast delivery may not be seen in operation.   However, the larger problem is that command handler methods and worker methods of `DAQModule` are in different threads and this is a problem waiting to bite module developers in general.

**Comment by glehmannmiotto** [comment]
I don’t really agree with this analysis. What you describe is only true if the commands are not implemented correctly. Assuming a worker thread is started at start and stopped at stop, then the start command shall not finish until the worker is actually active and the stop shall not complete until the thread has been terminated. Even in case of a fast delivery commands are strictly sequential.

**Comment by mroda88** [comment]
Would it be a good idea to add in the documentation that the implementation of the commands should be done in such a way that the command returns only when the module is able to accept new commands?

**Comment by brettviren** [comment]
All I know is I was able to observe a failed race.  This was about a month ago and I should have made the issue when it was fresh.  I was reminded of this issue recently so I figured better late than never.

If I look at `ThreadHelper`, probably what my tests saw was this throw

https://github.com/DUNE-DAQ/appfwk/blob/develop/include/appfwk/ThreadHelper.hpp#L123

But, maybe this is considered correct behavior.

start/stop are just two commands.

As we add other commands (eg a "status" or expert commands), the user is left to figure out how to write code to span these two threads.  It seems to me a "framework" should not inflict this on user code.  Right?

**Comment by philiprodrigues** [comment]
@brettviren do you remember any more about the steps that led to the failed race? I tried delivering fast repeated start/stops to `fdpc-job.json` with `fileCommandFacility` like this:

```
for i in {0..100}; do
    jq '.[2]' fdpc-job.json > /tmp/rodrigues/job.fifo 
    jq '.[3]' fdpc-job.json > /tmp/rodrigues/job.fifo
done
```

but no crashes

**Comment by brettviren** [comment]
The test consumed a sequence of commands direct from file.

```
daq_application --commandFacility file://fdpc-job.json
```

As before, rerunning this test now, I see no exception thrown and the `do_work()` methods are never run.  I think is correct behavior.  I retried the `listrev` test and also get this behavior.

So, like before, I can't reproduce the problem where the 30s hang on a queue timeout followed by crash occurred.

Looking at `ThreadHelper`, I think the worker thread should be protected from (and by) the handling of the **start** and **stop** commands.  And likewise from **init** and **conf** as long as they come before **start** and **stop**.  Understanding that code better now, I am unable to explain how the previous race occurred.  I can't now rule out my previous observation was due to something else than the `ThreadHelper` handling.

So, I have two conclusions:

1. To safely handle a command which arrives between the **start** and **stop** commands and for which the handling accesses data shared with the worker thread will require some form of protection.  Eg, a mutex around the shared data or a thread-safe transfer of the command from main to worker thread.  The former is likely slower than the latter.  The latter could be ensconced into the base class by formalizing into some kind of "worker pattern".

2. The nominal thread handling is still exposed to user code.  The **start** and **stop** handler methods and the registering of the `do_work()` method with a `ThreadHelper` is a pattern that will be common to all modules that do asynchronous work.  This code is a candidate for inclusion into the `DAQModule` base class or possibly a middle class layer like `DAQWorker`.  Such code lifting should be done in a way that still allows some modules to have zero or multiple threads.
